src/controller/python/chip-device-ctrl.py

```diff
@@ -385,6 +386,27 @@ def do_connect(self, line):
             print(str(ex))
             return
 
+    def do_resolve(self, line):
+        """
+        resolve <fabricid> <nodeid>
+
+        Resolve DNS-SD name corresponding with the given fabric and node IDs and
+        update address of the node in the device controller.
+        """
+        try:
+            args = shlex.split(line)
+            if len(args) == 2:
+                err = self.devCtrl.ResolveNode(int(args[0]), int(args[1]))
+                if err == 0:
+                    address = self.devCtrl.GetAddressAndPort(int(args[1]))
```

Comment:
Is the resolution synchronous? I don't find it easy to follow the entire path, however it seems that at least avahi is asyn: you ask for a resolve and will later get a callback of 'node id resolved'. 

What guarantees that GetAddressAndPort here is available as soon as ResolveNode completes?

Reply:
`ResolveNode` like other long-running commands spawn the asynchronous node id resolution using `CallAsync` function which basically waits until a completion event is signaled. So the underlying work is done asynchronously and only `ResolveNode` waits till it completes.